spanner/spansql/parser.go

```diff
@@ -1861,12 +1869,51 @@ func (p *parser) parseSelectFrom() (SelectFrom, *parseError) {
 		return sf, nil
 	}
 
-	// TODO: consume "HASH"
-
 	sfj := SelectFromJoin{
 		Type: jt,
 		LHS:  sf,
 	}
+	setHint := func(k, v string) {
+		if sfj.Hints == nil {
+			sfj.Hints = make(map[string]string)
+		}
+		sfj.Hints[k] = v
+	}
+	if hashJoin {
+		setHint("JOIN_METHOD", "HASH_JOIN")
+	}
+
+	if p.eat("@") {
+		if err := p.expect("{"); err != nil {
+			return nil, err
+		}
+		for {
+			if p.sniff("}") {
+				break
+			}
+			tok := p.next()
+			if tok.err != nil {
+				return nil, tok.err
+			}
+			k := tok.value
+			if err := p.expect("="); err != nil {
+				return nil, err
+			}
+			tok = p.next()
+			if tok.err != nil {
+				return nil, tok.err
+			}
+			v := tok.value
+			setHint(k, v)
```

Comment:
nit: If `HASH JOIN` is already specified in the join clause, it is illegal to also specify it as a join hint as well, so ideally we should check whether both are being set here.

Source: https://cloud.google.com/spanner/docs/query-syntax#join-hints

Reply:
Yeah, there's a lot of strictness throughout this parser that isn't done. It might be worth some future work to tighten things up at some stage, but I'm sticking with just working towards supporting valid things and not putting too much effort into rejecting invalid things.